irohad/torii/processor/impl/transaction_processor_impl.cpp

```diff
@@ -15,103 +15,118 @@
+        auto h = hash(tx).to_string();
+        this->proposal_set_.erase(h);
+        this->candidate_set_.insert(h);
+        this->notifier_.get_subscriber().on_next(
+            std::make_shared<TransactionResponse>(TransactionResponse{
+                h,
+                Status::STATEFUL_VALIDATION_SUCCESS,
+            }));
+      };
+
+      static const auto notify_fail = [this](const auto &h) {
+        this->notifier_.get_subscriber().on_next(
+            std::make_shared<TransactionResponse>(
+                TransactionResponse{h, Status::STATEFUL_VALIDATION_FAILED}));
+      };
+
       // move commited txs from proposal to candidate map
-      pcs_->on_commit().subscribe([this](
-          rxcpp::observable<model::Block> blocks) {
+      pcs_->on_commit().subscribe([this](auto blocks) {
         blocks.subscribe(
             // on next..
-            [this](model::Block block) {
-              for (const auto &tx : block.transactions) {
-                if (this->proposal_set_.count(hash(tx).to_string())) {
-                  proposal_set_.erase(hash(tx).to_string());
-                  candidate_set_.insert(hash(tx).to_string());
-                  TransactionResponse response;
-                  response.tx_hash = hash(tx).to_string();
-                  response.current_status =
-                      model::TransactionResponse::STATEFUL_VALIDATION_SUCCESS;
-                  notifier_.get_subscriber().on_next(
-                      std::make_shared<model::TransactionResponse>(response));
-                }
-              }
+            [this](auto block) {
+              static const auto in_proposal = [this](const auto &tx) {
+                return this->proposal_set_.count(hash(tx).to_string());
+              };
+              boost::for_each(
+                  block.transactions | boost::adaptors::filtered(in_proposal),
+                  notify_success);
             },
             // on complete
             [this]() {
-              for (auto tx_hash : proposal_set_) {
-                TransactionResponse response;
-                response.tx_hash = tx_hash;
-                response.current_status =
-                    TransactionResponse::STATEFUL_VALIDATION_FAILED;
-                notifier_.get_subscriber().on_next(
-                    std::make_shared<model::TransactionResponse>(response));
-              }
-              proposal_set_.clear();
-
-              for (auto tx_hash : candidate_set_) {
-                TransactionResponse response;
-                response.tx_hash = tx_hash;
-                response.current_status = TransactionResponse::COMMITTED;
-                notifier_.get_subscriber().on_next(
-                    std::make_shared<model::TransactionResponse>(response));
-              }
-              candidate_set_.clear();
+              boost::for_each(
+                  boost::join(this->proposal_set_, this->candidate_set_),
+                  notify_fail);
```

Comment:
Notify fail provides only in the end observable. This is wrong behavior at all.

Reply:
It's on complete callback, what's wrong?